### PR TITLE
Added new titus-mount-bind command

### PR DESCRIPTION
**Pull request by solarkennedy**
This new comman makes it possible to add bind mounts from
host->container while a container is running!

This is going to be used for the future /mnt-shared feature to allow
sidecars to share mounts with each other.

### CODE REVIEW

File: mount/Makefile [hunk 1]
```diff
@@ -11,11 +11,15 @@ titus-mount-ceph: titus-mount-ceph.c scm_rights.c common.h
...
+install: titus-mount-nfs titus-mount-ceph titus-mount-block-device titus-mount-bind
+	sudo rsync -a titus-mount-nfs titus-mount-block-device titus-mount-ceph titus-mount-bind /apps/titus-executor/bin/
 
 clean:
-	rm -f titus-mount-nfs titus-mount-block-device titus-mount-ceph
+	rm -f titus-mount-nfs titus-mount-block-device titus-mount-ceph titus-mount-bind
 
 fmt:
 	clang-format -i *.c *.h
+	iwyu -I /usr/lib/gcc/x86_64-linux-gnu/9/include -I /usr/include -isystem /usr/lib/gcc/x86_64-linux-gnu/9/include-fixed/ -I include/ titus-mount-bind.c 2>&1  | fix_include --comments
```
Comment (sargun) [inline mount/Makefile:25]: Maybe add a comment on how to install iwyu?

File: mount/titus-mount-bind.c [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+#include <assert.h> // for assert
+#include <fcntl.h> // for open, openat, AT_FDCWD, AT_RECURSIVE, O_PATH
+#include <limits.h> // for PATH_MAX
+#include <linux/mount.h> // for MOVE_MOUNT_F_EMPTY_PATH, OPEN_TREE_CLOEXEC
+#include <sched.h> // for setns, CLONE_NEWNS
+#include <syscall.h> // for __NR_fsopen, __NR_move_mount, __NR_open_tree
+#include <unistd.h> // for syscall, pid_t
+
+// Only to make it easier to build on older kernels
+#  define AT_RECURSIVE		0x8000
```
Comment (sargun) [inline mount/titus-mount-bind.c:15]: You're going to want to replace this with an:

```c
#ifndef AT_RECURSIVE
#define AT_RECURSIVE 0x8000
#endif
```

File: mount/titus-mount-bind.c [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+static inline int open_tree(int dfd, const char *filename, unsigned int flags)
+{
+	return syscall(__NR_open_tree, dfd, filename, flags);
+}
+
+static void switch_into_mount_namespace(int nsfd)
+{
+	int ret;
+	int mnt_fd = openat(nsfd, "mnt", O_RDONLY);
+	assert(mnt_fd != -1);
```
Comment (sargun) [inline mount/titus-mount-bind.c:52]: Please put a newline between variable declarations and code.

File: mount/titus-mount-bind.c [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+	ret = setns(mnt_fd, CLONE_NEWNS);
+	if (ret == -1) {
```
Comment (sargun) [inline mount/titus-mount-bind.c:54]: Any reason you're not using the E() macro above?

File: mount/titus-mount-bind.c (hunk 3 above)
Comment (sargun) [reply mount/titus-mount-bind.c:52]: Also, any reason you're not using the E macro above?

File: mount/titus-mount-bind.c [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+	 * We do this because parsing args is a bigger pain than passing
+	 * via environment variable, although passing via environment
+	 * variable has a "cost" in that they are limited in size
+	 */
+	const char *source = getenv("MOUNT_SOURCE");
+	char *target = getenv("MOUNT_TARGET");
+	const char *pid1dir = getenv("TITUS_PID_1_DIR");
+
+	errno = 0;
+	assert(errno == 0);
```
Comment (sargun) [inline mount/titus-mount-bind.c:73]: Maybe this is a dumb question, why do you set errno to 0?

File: mount/titus-mount-bind.c [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+	if (!(target && source && pid1dir)) {
+		fprintf(stderr,
+			"Usage: must provide MOUNT_SOURCE, MOUNT_TARGET, and TITUS_PID_1_DIR env vars");
```
Comment (sargun) [inline mount/titus-mount-bind.c:77]: newline at the end of this string.

File: mount/titus-mount-bind.c [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+		return 1;
+	}
+
+	/* This nsfd is used to extract other namespaces to switch to, similar to a pidfd */
+	char pid1_ns_path[PATH_MAX];
```
Comment (sargun) [inline mount/titus-mount-bind.c:82]: *Ideally* keep all variable declarations at the top (even though that's a C89 thing)

**Review by sargun — CHANGES_REQUESTED** [review]

File: mount/titus-mount-bind.c (hunk 5 above)
Comment (solarkennedy) [reply mount/titus-mount-bind.c:73]: no idea, copy pasted

**Pull request merged**